DSP Ops tester - Take 3

ranj063 opened the pull request:
This PR adds the API for enabling testing the DSP ops in the driver for testing the FW boot flow, PM flow etc. It introduces a new kernel config to enable the testing debugfs interface. When enabled, normal audio operations will be disabled.

For now the following ops are supported for testing:

1. boot_fw : To boot a firmware file pass from userspace by setting the fw_filename and fw_path debugfs files
2. dsp_power_state: To set the DSP state. CUrrent only D3 is supported with this op. SUpport for other states will be added soon.
3. fw_state: To check the current fw boot state

ranj063 committed:
ASoC: SOF: core: Introduce a new debug flag for DSP ops testing

Introduce a new debug option to enable testing of DSP operations. With
this ioption enabled, normal audio operations will be disabled. The API
for the testing interface will be introduced in the follow up patches.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

ranj063 committed:
ASoC: SOF: debug: Add debugfs entries for testing DSP ops

Add debugfs entries for testing DSP ops. This patch only adds the
firmware filename and path entries to pass them from userspace. These
will be saved in the newly added field for saving the test profile in
struct snd_sof_dev. These will be used for testing DSP ops such as booting
firmware and setting the DSP power state in the following patches.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

ranj063 committed:
ASoC: SOF: Modify the signature of load_firmware op

Modify the signature of the load_firmware op to take the firmware
filename as an argument. The firmware filename should be include the
full path of the firmware file to be loaded. This is in preparation
to add a debug op that will be used to boot a firmware with the
filename passed from userspace.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

ranj063 committed:
ASoC: SOF: debug-dsp-ops: Add a new op to test firmware boot

Add a new op to test firmware boot. Make sure to the DSP is powered off
before attempting to boot a new firmware.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

ranj063 committed:
ASoC: SOF: debug_dsp_ops: Add op to test DSP power state changes

Add a new op to test DSP power state changes. For now only D3 is
supported. More states will be added in follow up patches.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

ranj063 committed:
ASoC: SOF: debug-dsp-ops: Add a debugfs entry to read current FW status

Add a read-only debugfs entry to read the current FW boot state.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

plbossart commented on sound/soc/sof/Kconfig:
```diff
@@ -266,6 +266,15 @@ config SND_SOC_SOF_DEBUG_RETAIN_DSP_CONTEXT
+	  operations will be disabled when this option is selected.
+	  Say Y if you want to enable DSP ops testing.
+	  If unsure, select "N".
+
```
We should make it a SND_SOC_SOF_DEBUG_DSP_OPS_TEST_ENABLE, to let distros opt-out, and then set the actual boolean with a kernel parameter.

it's too much work for everyone to have yet another build.

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -452,6 +452,15 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
+	 * skip loading/booting firmware and registering the machine driver when DSP OPS testing
+	 * is enabled. Normal audio operations will be unavailable in this mode.
+	 */
+	if (IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_DSP_OPS_TEST)) {
```
kernel parameter

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -588,7 +597,9 @@ int snd_sof_device_probe(struct device *dev, struct snd_sof_pdata *plat_data)
-	if (sof_debug_check_flag(SOF_DBG_DSPLESS_MODE)) {
+	/* select DSPless mode only when DSP OPS testing is not in use */
+	if (sof_debug_check_flag(SOF_DBG_DSPLESS_MODE) &&
+	    !IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_DSP_OPS_TEST)) {
```
I would put this DSP_OPS_TEST as a new bitfield in sof_debug

ujfalusi replied:
Document the DSPless VS ops_test mode and print if the DSPless mode is ignored due to the ops_test taking precedence.
Or should DSPless mode be the one which should have the higher priority?

plbossart commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: GPL-2.0-only
+//
+// Copyright(c) 2024 Intel Corporation. All rights reserved.
```
remove All rights reserved.

plbossart commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -8,6 +8,97 @@
+		snd_sof_fw_unload(sdev);
+
+	/* free trace */
+	sof_fw_trace_free(sdev);
```
this doesn't seem related to the boot but more a shutdown...

plbossart commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -8,6 +8,97 @@
+	sof_fw_trace_free(sdev);
+
+	/* power up the DSP */
+	ret = snd_sof_dsp_runtime_resume(sdev);
```
Yikes. power up the DSP and dsp_runtime_resume doesn't seem like the same thing at all to me.

plbossart commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -115,5 +209,9 @@ int sof_dbg_dsp_ops_test_init(struct snd_sof_dev *sdev)
+	if (ret < 0)
+		return ret;
+
+	return sof_dsp_dsp_ops_create_dfse(sdev, "boot_fw", dsp_ops_debugfs, 0222);
```
No control on whether we boot from IMR?

ranj063 replied:
Always a fresh boot with the boot option. I will need to add another option to do IMR boot later

plbossart replied:
The IMR boot is the target, I don't think anyone is interested in a fresh boot, historically this created all kinds of issues in stress tests.

ranj063 replied:
ok, fixed now to do IMR boot as default

plbossart commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -213,5 +245,9 @@ int sof_dbg_dsp_ops_test_init(struct snd_sof_dev *sdev)
+	if (ret < 0)
+		return ret;
+
+	return sof_dsp_dsp_ops_create_dfse(sdev, "dsp_power_state", dsp_ops_debugfs, 0666);
```
The problem I have with this API is that it's not quite clear how a script would know that the dsp_power_state has been reached.
How long should the script wait and does it get any error codes?

This seems to simple to be useful IMHO.

ranj063 replied:
@plbossart i have exposed both the current fw boot state and the dsp power state that the script can poll to see if the desired state has been reached. How long to poll something that we can set based on the current kernel poll wait times. As for error codes, Im afraid I have no way to pass that on to userspace but we have everything we need in the dmesg isnt it?

plbossart replied:
Well, it's rather problematic if you don't have an error code for an API, and rely on something else to find if the routine worked or not... If you start fuzzing or doing all kinds of stress tests, you may have difficulties finding out when exactly the problem occured. it's just very hard to parse such messages. Exhibit A is the SOF CI, where we need to have a start marker and we stop on dev_err but that leaves lots of issues undetected...

It's also problematic if we have to duplicate in scripts the timeouts we already use for the kernel.

I do think we ought to work first on the design/architecture before going into the code review space. It's a very large effort we are starting with scripts that will have to be maintained. We need to do this right from the beginning.

plbossart reviewed (commented):
I am not on-board with the directions, the simple "echo 1 > boot" or echo 3 > dsp_state is too simple for scripting. These transitions take time, so there's no way for a script to know how long it needs to wait and if the transition was successful.

I would recommend writing a 2-page summary of what we want to achieve, requirements included, otherwise we'll have lots of code review without first agreeing on what we want to achieve.

ranj063 committed:
ASoC: SOF: debug-dsp-ops: Add op for unloading firmware

Unload any existing firmware and set the fw_state to PREPARE in
preparation for a subsequent firmware boot.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

ranj063 committed:
ASoC: SOF: debug-dsp-ops: Add ops for trace init/free

These will be used to debugging purposes when testing the different DSP
ops.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

ranj063 committed:
ASoC: SOF: Set initial DSP device state to D3

Before the first firmware boot, set the initial state of the DSP to D3.
The DSP D3 state is not the same as the SOF PCI device D3. Instead, it
represents that the DSP is in the powered off state.

After firmware boot is completed successfully, set the state to D0 to
indicate that the DSP has been powered on.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

ranj063 committed:
ASoC: SOF: Intel: hda: Move DSP power step to the set_power_state op

This will allow for the set_power_state op to be used for switching the
DSP to the D3 state. The logic for powering up the DSP when setting the
state to D0 also needs to be moved and will be done in a follow up. But
for now, this will allow stress testing firmware boot.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

plbossart commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -63,6 +63,14 @@ static int sof_dsp_ops_set_power_state(struct snd_sof_dev *sdev, char *state)
+static void sof_dsp_ops_unload_firmware(struct snd_sof_dev *sdev)
+{
+	if (sdev->basefw.fw)
+		snd_sof_fw_unload(sdev);
```
Not sure what this does? Is this actually a capability?

ranj063 replied:
there's no unloading the firmware from the DSP but I need to reset the sdev->base_fw to be able to start from a clean state for the next iteration

plbossart replied:
a comment would help then, things might be different on other platforms.

plbossart reviewed (commented):
There are multiple comments not addressed, not sure if they were missed or github messes with the comments.
The ordering of the patches is weird, we should first have code moves then the addition of new functionality.

ujfalusi commented:
> I am not on-board with the directions, the simple "echo 1 > boot" or echo 3 > dsp_state is too simple for scripting. These transitions take time, so there's no way for a script to know how long it needs to wait and if the transition was successful.

to note:
boot: `echo on > "/sys/bus/pci/devices/0000:00:1f.3/power/control"`
power off: `echo auto > "/sys/bus/pci/devices/0000:00:1f.3/power/control"`

> I would recommend writing a 2-page summary of what we want to achieve, requirements included, otherwise we'll have lots of code review without first agreeing on what we want to achieve.

ujfalusi commented on sound/soc/sof/debug.c:
```diff
@@ -371,9 +371,12 @@ int snd_sof_dbg_init(struct snd_sof_dev *sdev)
-					sizeof(sdev->fw_state),
-					"fw_state", 0444);
+	err = snd_sof_debugfs_buf_item(sdev, &sdev->fw_state, sizeof(sdev->fw_state), "fw_state",
+				       0444);
```
nitpick: it would look better if the `"fw_state"` is moved to the second line.

ujfalusi commented on sound/soc/sof/amd/acp-loader.c:
```diff
@@ -263,18 +263,19 @@ int acp_sof_dsp_run(struct snd_sof_dev *sdev)
 EXPORT_SYMBOL_NS(acp_sof_dsp_run, SND_SOC_SOF_AMD_COMMON);
 
-int acp_sof_load_signed_firmware(struct snd_sof_dev *sdev)
+int acp_sof_load_signed_firmware(struct snd_sof_dev *sdev, const char *fw_filename)
```
for the commit message: `fw_filename` is not just the filename, but the path+filename...

ujfalusi commented on sound/soc/sof/amd/acp-loader.c:
```diff
@@ -263,18 +263,19 @@ int acp_sof_dsp_run(struct snd_sof_dev *sdev)
+	if (!fw_filename) {
+		fw_filename = kasprintf(GFP_KERNEL, "%s/%s",
+					plat_data->fw_filename_prefix,
+					adata->fw_code_bin);
```
What about the `fw_data_bin`? acp needs two files to be loaded...

ranj063 replied:
ahh I missed this. I don't know how to handle this for ACP with 2 files. Let me fix the ACP loader to ignore the pass argument for now

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -8,6 +8,33 @@
 #include "sof-priv.h"
 #include "ops.h"
 
```
typo in commit message "ddebug"

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -49,6 +76,21 @@ static ssize_t sof_dsp_ops_tester_dfs_write(struct file *file, const char __user
+			return -ENOMEM;
+
+		size = simple_write_to_buffer(string, count, ppos, buffer, count);
+		kfree(string);
```
so you can write anything and it will boot a firmware?

ranj063 replied:
as long as it is a valid file, we will attempt to parse it and if it is a valid firmware, yes we will boot it

ujfalusi replied:
I can `echo boo > /sys/kernel/debug/sof/dsp_ops/boot_fw` and will boot? Hehe

ranj063 replied:
Yeah I guess thats true and in some sense I feel like its alright. I can add a check here but what do I check for? echo 1 or echo true or echo go? It shouldnt really matter right? as long as we're writing something to boot_fw, it all signals a go

ujfalusi replied:
It is fine, just fun.

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -8,6 +8,33 @@
+
+	sdev->first_boot = true;
+
+	if (!sdev->test_profile.fw_path || !sdev->test_profile.fw_name) {
```
Why don't we set these to the currently used path and firmware initially?
Or use the currently loaded firmware to boot?

ranj063 replied:
I dont want to use the normal FW profile used on the DUT for regular audio functionality. They should be read only so that the user cannot modify it randomly

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -8,6 +8,33 @@
+	const char *fw_filename;
+	int ret;
+
+	sdev->first_boot = true;
```
you should be setting this right before you call `snd_sof_load_firmware()`

ujfalusi replied:
You should really be setting this before starting the load, we have returns in-between now.

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -8,6 +8,33 @@
+				sdev->test_profile.fw_name);
+
+	/* load firmware */
+	ret = snd_sof_load_firmware(sdev, fw_filename);
```
How this will pass our test in `snd_sof_load_firmware_raw()` ?
```c
	/* Don't request firmware again if firmware is already requested */
	if (sdev->basefw.fw)
		return 0;
```
You likely need to release the firmware, wipe all information about modules, possibly drop loaded external libraries as well.
Likely the topology needs to be re-verified to see if the firmware have all the modules that is needed, so destroy the card?

ranj063 replied:
see my comment below. With the debug option selected nothing is loaded at boot

ujfalusi replied:
right, but you can load it once, is this by design?
And on consecutive tries no one would know if the new firmware was ignored.

ranj063 replied:
The problem is if someone wants to load a new firmware, they need to make sure the DSP is powered off first and the previous FW is unloaded before attempting to boot again. Im not sure we should handle these in the kernel.

ujfalusi replied:
Should we tell the user that what is tried is going to be a NOP and the previously loaded kernel is going to remain?

plbossart replied:
We need to be clear on what we mean by 'new firmware'. if someone overrides an existing firmware in /lib/firmware, they will face all kinds of caching issues. We cannot really add a new firmware with SSH and hope that it's used immediately, that's really asking for trouble.

The assumption should be that you test the firmware that was available at boot.

marc-hb replied:
> The assumption should be that you test the firmware that was available at boot.

"Only at boot" and "all kinds of caching issues" sound a bit too extreme to me. I mean the requirement to reboot to test a different firmware version sounds too extreme? I vaguely remember some `request_firmware()` mis-use and caching bug a couple years ago but besides that one, https://github.com/thesofproject/sof-test/blob/main/tools/kmod/sof_remove.sh has been working just fine to test new firmware. Granted this was before libraries but  why couldn't this achieve the same? At least when libraries are not used.

I this requires some mass `rmmod` like `sof_remove.sh` does and like @ujfalusi seems to request then it looks good to me.

BTW do we care about libraries here? I thought this was meant to debug early DSP boot issues. Once those are fixed, just debug the firmware as usual/before?

cc: @andyross

marc-hb replied:
> Should we tell the user that what is tried is going to be a NOP and the previously loaded kernel is going to remain?

NEVER, ever silently discard user input. That's sending people who are already losing hair debugging problems straight to the asylum. If you must discard user input for some reason, then it's fine but return every major error code you can (for scripts), and log at the highest level every error message you can - in bold face and capitals (for both scripts and humans)

ranj063 replied:
OK I've fixed this to boot default FW is the user chooses not to pass the test firmware file

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -8,6 +8,33 @@
+	/* boot firmware */
+	sof_set_fw_state(sdev, SOF_FW_BOOT_IN_PROGRESS);
+
+	return snd_sof_run_firmware(sdev);
```
I think this is going to be be really dangerous. you have records of the initially booted firmware, the modules of it and then you load a different firmware where the modules might be in different places or not even exists?

ranj063 replied:
there's no record of any initially loaded firmware. When the debug option is selected, i block everything during probe including loading the default firmware

ujfalusi replied:
but again, you will have records of the first loaded firmware and it's libraries.

ujfalusi commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -149,13 +149,21 @@ static size_t sof_ipc4_fw_parse_basefw_ext_man(struct snd_sof_dev *sdev)
 
 	payload_offset = sof_ipc4_fw_parse_ext_man(sdev, fw_lib);
+	if (IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_DSP_OPS_TEST)) {
+		kfree(fw_lib);
```
you can use `devm_kfree(sdev->dev, fw_lib);`

ranj063 replied:
fixed now.

ujfalusi commented on sound/soc/sof/loader.c:
```diff
@@ -170,6 +170,9 @@ int snd_sof_run_firmware(struct snd_sof_dev *sdev)
 	sof_set_fw_state(sdev, SOF_FW_BOOT_COMPLETE);
+	
+	if (IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_DSP_OPS_TEST))
+		return 0;
```
I think we should not exclude the `snd_sof_dsp_post_fw_run()`

ranj063 replied:
post_fw_run() does things that we dont care about when just testing fw boot like for example in the MTL we do sdw startup &  enable sdw interrupts. We dont want all those.

ujfalusi commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -149,13 +149,21 @@ static size_t sof_ipc4_fw_parse_basefw_ext_man(struct snd_sof_dev *sdev)
+	if (IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_DSP_OPS_TEST))
+		fw_lib = kzalloc(sizeof(*fw_lib), GFP_KERNEL);
+	else
+		fw_lib = devm_kzalloc(sdev->dev, sizeof(*fw_lib), GFP_KERNEL);
```
have you checked what happens with IPC3?

ranj063 replied:
when I'm doing a 10000 times firmware boot , do you think it still makes sense?

ujfalusi replied:
On module remove you are going to leak these, we rely on devm to handle these (see `sof_ipc4_exit()`)

ujfalusi replied:
As for IPC3, I was merely asking if you have checked how this works with ipc_type=0.

ranj063 replied:
I made the test mode optional for use only with IPC4

ujfalusi replied:
you are still leaking the `fw_lib->modules`

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -8,6 +8,61 @@
+
+			/*
+			 * Fix set_power_state to power up the DSP.
+			 * ATM this is taken care of during firmware boot
```
what is the meaning of this comment?
If something needs to be fixed, then why aren't you fixing it?

ujfalusi replied:
I still don't get the meaning of this comment....

ranj063 replied:
today we handle powering up the DSP during snd_sof_run_firmware(). So, if you want to go from D3 -> D0, we have to write to boot_fw() and it will handle it automatically. I want to fix this later on to split the DSP powering up out of the run_firmware call. But this is not critical right now

ujfalusi replied:
but the power up is a bit more than just toggle a switch, we need to 'send IPC' (the boot config needs to be placed), we need to wait for IPC (response that the DSP is ready to either start the download or it booted with IMR.

Is it worth adding confusing comments for something that will happen at some point. Or not.

plbossart replied:
I was also unable to understand the comment.

ranj063 replied:
ok I've removed support for setting DSP to D0 out for now. This is handled automatically during load_firmware for now.

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -8,6 +8,61 @@
+			 * Fix set_power_state to power up the DSP.
+			 * ATM this is taken care of during firmware boot
+			 */
+			return snd_sof_dsp_set_power_state(sdev, &target_state);
```
Would a `pm_runtime_get_sync()` boot the DSP up?

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -8,6 +8,61 @@
+			return snd_sof_dsp_set_power_state(sdev, &target_state);
+		}
+
+		return 0;
```
IPC3 might need `ctx_restore`?

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -252,5 +284,9 @@ int sof_dbg_dsp_ops_test_init(struct snd_sof_dev *sdev)
+	if (ret < 0)
+		return ret;
+
+	return sof_dsp_dsp_ops_create_dfse(sdev, "fw_state", dsp_ops_debugfs, 0444);
```
fwiw, we are exposing the `fw_state` as a read-only buffer.
Which is not too user friendly, I agree.

ranj063 replied:
making it read-write would be too confusing with all the stages we have for fw boot isnt it?

ujfalusi replied:
No, what I mean is that we have this information already, but it is just returning the raw u32.
I don't know if it is used by anyone.

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -175,6 +183,17 @@ static ssize_t sof_dsp_ops_tester_dfs_write(struct file *file, const char __user
+
+		size = simple_write_to_buffer(string, count, ppos, buffer, count);
+		kfree(string);
+		sof_dsp_ops_unload_firmware(sdev);
```
you would need to invalidate all information related to the loaded firmware, like modules and stuff...

ranj063 replied:
Typically unload_firmware would be called after the DSP is put into D3 so I dont think we need to worry about modules here yet. As for additional libraries, yes I will need to handle it when  I add support for stress testing library support in a follow up.

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -311,5 +327,13 @@ int sof_dbg_dsp_ops_test_init(struct snd_sof_dev *sdev)
+	if (ret < 0)
+		return ret;
+
+	return sof_dsp_dsp_ops_create_dfse(sdev, "free_trace", dsp_ops_debugfs, 0222);
```
how do you start the trace after powering on the DSP? Usually we need to send IPC messages. Does the `sof_resume()` will be called to take care of it?

ranj063 replied:
Our regular sequence for starting trace after boot is: run_firmware() -> trace_init(). 

So, in the test script, we'd do the same. boot the firmware and then call init_trace. And then do whatever else we want to do say for example load libaries, topologies etc

ujfalusi replied:
Can you share the test script? It is hard to visualize how this is used.

marc-hb replied:
Oooohh, so that's why you haven't been reviewing the test script @ujfalusi ? Mere lack of marketing? The PR is here:
- https://github.com/thesofproject/sof-test/pull/1180

ujfalusi commented on sound/soc/sof/core.c:
```diff
@@ -418,8 +418,14 @@ static int sof_init_environment(struct snd_sof_dev *sdev)
 
+	/* set initial power state to D3 */
+	sdev->dsp_power_state = state;
+
```
Why we got away without this before?
I'm not sure what is the issue....

ranj063 replied:
we didnt get away, it wasn't problematic because we never did multiple firmware boots while the PCI device is in D0. So for us the DSP D0 and PCI device D0 always meant the same. But if we were to boot the firmware multiple times without going through a suspend cycle, I need to put the DSP is D3 before I load & boot the firmware multiple times

ujfalusi replied:
in `snd_sof_device_probe() we
```c
	/* initialize default DSP power state */
	sdev->dsp_power_state.state = SOF_DSP_PM_D0;
```
and then later in here we:
```c

	/* set initial power state to D3 */
	sdev->dsp_power_state = state;
```
Which is and from where right?

ranj063 replied:
good catch. the first initialize to default power state is incorrect.

ujfalusi commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -1059,12 +1062,10 @@ int hda_dsp_suspend(struct snd_sof_dev *sdev, u32 target_state)
 	ret = hda_suspend(sdev, false);
-	if (ret < 0) {
+	if (ret < 0)
 		dev_err(bus->dev, "error: suspending dsp\n");
```
while here, can we drop the `error:` from the print (and print out the function name)?

ujfalusi commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -684,8 +684,22 @@ static int hda_dsp_set_power_state(struct snd_sof_dev *sdev,
+
+			ret = chip->power_down_dsp(sdev);
+			if (ret < 0) {
+				dev_err(sdev->dev, "failed to power down DSP during suspend\n");
```
"%s: failed to power down DSP\n" ?

ujfalusi reviewed (commented):
@ranj063, I'm still not sure about the use of this.
We will introduce a new flow which will be tested and it is different than what we normally use, and we draw conclusion based on the test results?

I also think that if this mode is selected then we should not load topology, not create sound card. I see this as a really dangerous feature which can introduce unexpected side-effects on the normal use of the stack.

ranj063 commented:
> @ranj063, I'm still not sure about the use of this. We will introduce a new flow which will be tested and it is different than what we normally use, and we draw conclusion based on the test results?
> 
> I also think that if this mode is selected then we should not load topology, not create sound card. I see this as a really dangerous feature which can introduce unexpected side-effects on the normal use of the stack.

Yes, agreed. Thats what the mode does. There's no audio functionality when this mode is selected

ujfalusi commented on sound/soc/sof/sof-priv.h:
```diff
@@ -228,7 +228,7 @@ struct snd_sof_dsp_ops {
 
 	/* FW loading */
-	int (*load_firmware)(struct snd_sof_dev *sof_dev); /* mandatory */
+	int (*load_firmware)(struct snd_sof_dev *sof_dev, const char *fw_filename); /* mandatory */
```
It should be documented that the fw_filename must contain the full path, including the filename.

ujfalusi replied:
Why have you dismissed this comment without nacking or addressing it?

ranj063 replied:
it is addressed in ops.h where the op is implemented

ujfalusi commented on sound/soc/sof/core.c:
```diff
@@ -418,8 +418,14 @@ static int sof_init_environment(struct snd_sof_dev *sdev)
 static int sof_probe_continue(struct snd_sof_dev *sdev)
 {
 	struct snd_sof_pdata *plat_data = sdev->pdata;
+	struct sof_dsp_power_state state = {
```
The 'const police' will say that this should be 'const'?

ranj063 replied:
do you think? I'm not passing it to another function here. So I guess it should be OK right?

ujfalusi replied:
They struck at random ;)
no need to do anything imho either.

ujfalusi commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -684,8 +684,23 @@ static int hda_dsp_set_power_state(struct snd_sof_dev *sdev,
+
+			ret = chip->power_down_dsp(sdev);
+			if (ret < 0) {
+				dev_err(sdev->dev, "%s: failed to power down DSP during suspend\n",
```
the 'during suspend' is redundant, no?
Is it even correct in case we remove the drivers and the power off fails?

ranj063 replied:
sorry not following the "is it correct?" comment? Are you asking if this is correct to call this during driver remove or are you asking is it OK to fail?

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -0,0 +1,119 @@
+	if (!strcmp(dentry->d_name.name, "fw_filename"))
+		string = sdev->test_profile.fw_name;
+	else if (!strcmp(dentry->d_name.name, "fw_path"))
+		string = sdev->test_profile.fw_path;
```
You are planning to extend this to libraries and topologies? Libraries likely.
Just thinking if we could get away with a single 'firmware' string which includes the path and the name.

ranj063 replied:
Yes libraries and topologies both. But as far as the topology is concerned, the test_profile can also be used to set the topology filename and path to test. I'll have to think about the library test when I get to it later

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -8,6 +8,33 @@
+	sdev->first_boot = true;
+
+	if (!sdev->test_profile.fw_path || !sdev->test_profile.fw_name) {
+		dev_dbg(sdev->dev, "Invalid firmware path or filename\n");
```
well, it is not invalid, but either of them is missing. But which one? This would be asked by the user (who might not even see the error since it is in debug level...)

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -0,0 +1,119 @@
+	int ret;
+
+	/* debugfs root directory for DSP ops debug */
+	dsp_ops_debugfs = debugfs_create_dir("dsp_ops", sdev->debugfs_root);
```
I cannot put the 'dsp_ops' to anywhere to be honest, strange name.

ranj063 replied:
what would you suggest I call it instead of "dsp_ops"?

ujfalusi replied:
"dsp_test" | "dsp_test_mode" | "test_mode" ?
Probably the "test_mode" is better as it is not really testing the DSP as such.

But I'm bad at naming.

marc-hb replied:
This will hopefully be used to find more bugs in the firmware/software than bugs in the DSP hardware! So how about `fw_debug_ops`? Then you can all the now redundant `fw_` prefixes in `fw_debug_ops/fw_*` below.

ranj063 replied:
ok renamed to fw_debug_ops. But I will leave the fw_ prefix for now because there will also be tplg_filename/ path etc.

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -0,0 +1,119 @@
+	return count;
+}
+
+
```
extra empty line

plbossart replied:
still there.

ujfalusi commented on sound/soc/sof/core.c:
```diff
@@ -582,17 +599,16 @@ int snd_sof_device_probe(struct device *dev, struct snd_sof_pdata *plat_data)
-	if (sof_debug_check_flag(SOF_DBG_DSPLESS_MODE)) {
+	/* select DSPless mode only when DSP OPS testing is not in use */
+	if (sof_debug_check_flag(SOF_DBG_DSPLESS_MODE) &&
+	    !sof_debug_check_flag(SOF_DBG_DSP_OPS_TEST_MODE)) {
```
I would still inform the user that the DSPless mode is ignored because a DSP test mode is selected and taken precedence.

plbossart replied:
i would really flip the logic and only take the DSP test mode when the user didn't explicitly bypass the DSP.

ujfalusi commented on sound/soc/sof/sof-priv.h:
```diff
@@ -52,6 +52,9 @@ struct snd_sof_pcm_stream;
 							 */
 #define SOF_DBG_DSPLESS_MODE			BIT(15) /* Do not initialize and use the DSP */
+#define SOF_DBG_DSP_OPS_TEST_MODE		BIT(16) /* Used for DSP ops testing.
+							 * No audio functionality
```
and overrides the DSPless mode.

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -120,6 +120,38 @@ static ssize_t sof_dsp_ops_tester_dfs_read(struct file *file, char __user *buffe
+			string = "PREPARE\n";
+			break;
+		case SOF_FW_BOOT_IN_PROGRESS:
+			string = "IN PROGRESS\n";
```
What is "In progress"?

ujfalusi commented on sound/soc/sof/debug-dsp-ops.c:
```diff
@@ -120,6 +120,38 @@ static ssize_t sof_dsp_ops_tester_dfs_read(struct file *file, char __user *buffe
+			string = "READY OK\n";
+			break;
+		case SOF_FW_BOOT_COMPLETE:
+			string = "COMPLETE\n";
```
I kind of like the numeric values as well (got used to them in the kernel log)...
Not sure if it makes sense to include it in the print (`6: BOOT_COMPLETE`)

ranj063 replied:
done

ujfalusi reviewed (commented):
@ranj063, it looks much cleaner now for sure.

I still have reservation on the usefulness of such interface. We test isolated and narrowed down cases which will never going to be possible to use in real use.

Probably it is beneficial, and I just cannot see that. The added code and complexity is minimal, which is a good sign.

ujfalusi commented on sound/soc/sof/ops.h:
```diff
@@ -474,12 +474,12 @@ snd_sof_pcm_platform_trigger(struct snd_sof_dev *sdev,
 
-/* Firmware loading */
-static inline int snd_sof_load_firmware(struct snd_sof_dev *sdev)
+/* Firmware loading. fw_filename must contain the full path to the file */
```
I see, but then it should also say that it can be left to NULL to use the strings from the default profile.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -565,6 +568,12 @@ struct snd_sof_dev {
 
+	/*
+	 * Flag set when DSP ops testing is enabled. When this is set normal audio functionality
+	 * including DSPless mode is disabled.
```
I really believe this is only relevant IF the DSPless mode is NOT enabled. It's more logical to me...

ranj063 replied:
ok I've flipped the logic to have DSPLESS mode take priority over ops testing

marc-hb commented:
(can't reply to the original anymore because Github does not like force pushes)

> The IMR boot is the target, I don't think anyone is interested in a fresh boot, historically this created all kinds of issues in stress tests.

I don't understand this. AFAIK, the purpose of this PR is to debug boot issues. We now booting from IMR versus not have had different sets of issues in the past. So shouldn't this PR allow testing both? I don't known which default would be best but I feel like both should be possible.

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -462,6 +462,17 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
+		sdev->probe_completed = true;
+		return 0;
+	}
+
```
typo in commit message? 
"idebug"

plbossart reviewed (dismissed):
Looks ok to me, I am just not sure what the next steps are?

ranj063 commented:
> Looks ok to me, I am just not sure what the next steps are?

@plbossart the first tiny next step is to have the fw boot script run with our CI to do the stress test. I also want to keep building the test script library to add more things like PM flows, DMA flows, topology loading, IPC etc

ranj063 commented:
@plbossart just a minor change to add the fw filename and fw_path string lengths

ranj063 commented:
@ujfalusi any objection to merging this? This will help me build upon it and add more test cases.
